src/main/java/com/microsoft/azure/batch/TaskOperations.java

```diff
@@ -118,61 +120,94 @@ public Exception getException() {
             return this.exception;
         }
 
+		/**
+		 * Submits one chunk of tasks to a job.
+		 * 
+		 * @param taskList A list of {@link TaskAddParameter tasks} to add.
+		 */
+        private void submit_chunk(List<TaskAddParameter> taskList) {
+        	// The option should be different to every server calls (for example,
+        	// client-request-id)
+        	TaskAddCollectionOptions options = new TaskAddCollectionOptions();
+        	this.bhMgr.applyRequestBehaviors(options);
+        	try {
+        		TaskAddCollectionResult response = this.client.protocolLayer().tasks().addCollection(this.jobId,
+        				taskList, options);
+
+        		if (response != null && response.value() != null) {
+        			for (TaskAddResult result : response.value()) {
+        				if (result.error() != null) {
+        					if (result.status() == TaskAddStatus.SERVER_ERROR) {
+        						// Server error will be retried
+        						for (TaskAddParameter addParameter : taskList) {
+        							if (addParameter.id().equals(result.taskId())) {
+        								pendingList.add(addParameter);
+        								break;
+        							}
+        						}
+        					} else if (result.status() == TaskAddStatus.CLIENT_ERROR
+        							&& !result.error().code().equals(BatchErrorCodeStrings.TaskExists)) {
+        						// Client error will be recorded
+        						failures.add(result);
+        					}
+        				}
+        			}
+        		}
+        	} catch (BatchErrorException e) {
+        		// If we get RequestBodyTooLarge could be that we chunked the tasks too large.
+        		// Try decreasing the size unless caused by 1 task.
+        		if (e.body().code().equals(BatchErrorCodeStrings.RequestBodyTooLarge) && taskList.size() > 1) {
+        			// Use binary reduction to decrease size of submitted chunks
+        			int midpoint = taskList.size() / 2;
+        			// If the midpoint is less than the currentMaxTasks used to create new chunks,
+        			// attempt to atomically reduce currentMaxTasks.
+        			// In the case where compareAndSet fails, that means that currentMaxTasks which
+        			// was the goal
+        			int max = currentMaxTasks.get();
+        			if (midpoint < max) {
+        				currentMaxTasks.compareAndSet(max, midpoint);
+        			}
+        			// Submit chunk in two equal sublists
+        			submit_chunk(taskList.subList(0, midpoint));
```

Comment:
You can submit the first half now, but push last half back to queue, so the second half may submit parallel with first half if there has free thread.